File: src/main/java/io/github/shirohoo/library/application/service/book/BookService.java
package io.github.shirohoo.library.application.service.book;

import io.github.shirohoo.library.application.domain.book.Book;
import io.github.shirohoo.library.application.domain.user.User;
import io.github.shirohoo.library.domain.book.BookRepository;
import io.github.shirohoo.library.domain.user.UserLoanHistoryRepository;
import io.github.shirohoo.library.domain.user.UserRepository;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
public class BookService {
    private final BookRepository bookRepository;
    private final UserRepository userRepository;
    private final UserLoanHistoryRepository userLoanHistoryRepository;

    public BookService(
            BookRepository bookRepository,
            UserRepository userRepository,
            UserLoanHistoryRepository userLoanHistoryRepository
    ) {
        this.bookRepository = bookRepository;
        this.userRepository = userRepository;
        this.userLoanHistoryRepository = userLoanHistoryRepository;
    }

    @Transactional
    public void saveBook(Book newBook) {
        if (bookRepository.save(newBook) == null) {
            throw new IllegalStateException("failed to save new book.");
        }
    }

    @Transactional
    public void loanBook(String username, String bookTitle) {
        if (userLoanHistoryRepository.existsBy(bookTitle, false)) {
            throw new IllegalArgumentException("this book has already been borrowed.");
        }

        User user = userRepository.findBy(username).orElseThrow();
        Book book = bookRepository.findBy(bookTitle).orElseThrow();

        user.loanBook(book);
    }

    @Transactional
    public void returnBook(String username, String bookTitle) {
        User user = userRepository.findBy(username).orElseThrow();
        user.returnBook(bookTitle);
    }
}

File: src/main/kotlin/io/github/shirohoo/library/application/service/book/BookService.kt
package io.github.shirohoo.library.application.service.book

import io.github.shirohoo.library.domain.book.Book
import io.github.shirohoo.library.domain.book.BookRepository
import io.github.shirohoo.library.domain.book.BookType
import io.github.shirohoo.library.domain.user.UserLoanHistoryRepository
import io.github.shirohoo.library.domain.user.UserLoanStatus
import io.github.shirohoo.library.domain.user.UserRepository
import kotlinx.coroutines.async
import kotlinx.coroutines.runBlocking
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
class BookService(
    private val bookRepository: BookRepository,
    private val userRepository: UserRepository,
    private val userLoanHistoryRepository: UserLoanHistoryRepository,
) {
    @Transactional
    fun saveBook(newBook: Book) = bookRepository.save(newBook)

    @Transactional
    fun loanBook(username: String, bookTitle: String) {
        if (userLoanHistoryRepository.existsBy(bookTitle, UserLoanStatus.LOANED)) {
            throw IllegalArgumentException("this book has already been borrowed.")
        }

        runBlocking {
            val user = async { userRepository.findBy(username) }
            val book = async { bookRepository.findBy(bookTitle) }

            user.await().loanBook(book.await())
        }
    }

    @Transactional
    fun returnBook(username: String, bookTitle: String) {
        val user = userRepository.findBy(username)
        user.returnBook(bookTitle)
    }

    @Transactional(readOnly = true)
    fun countLoanedBooks(): Int {
        return userLoanHistoryRepository.findAllByStatus(UserLoanStatus.LOANED)
            .count()
    }

    @Transactional(readOnly = true)
    fun getBookStatistics(): Map<BookType, List<Book>> {
        return bookRepository.findAll()
            .groupBy(Book::type)
    }
}
